streaming/src/main/scala/org/apache/gearpump/streaming/refactor/state/StateSpecs.scala

```diff
@@ -0,0 +1,208 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gearpump.streaming.refactor.state
+
+import java.util.Objects
+
+import org.apache.gearpump.streaming.refactor.coder.Coder
+import org.apache.gearpump.streaming.refactor.state.api.{BagState, MapState, SetState, ValueState}
+
+/**
+ * a utility object for StateSpec
+ */
+object StateSpecs {
+
+  private class ValueStateSpec[T](coder: Coder[T]) extends StateSpec[ValueState[T]] {
+
+    var aCoder: Coder[T] = coder
+
+    override def bind(id: String, binder: StateBinder): ValueState[T] = {
+      binder.bindValue(id, this, aCoder)
+    }
+
+    override def offerCoders(coders: Array[Coder[ValueState[T]]]): Unit = {
+      if (this.aCoder == null) {
+        if (coders(0) != null) {
+          this.aCoder = coders(0).asInstanceOf[Coder[T]]
+        }
+      }
+    }
+
+    override def finishSpecifying: Unit = {
+      if (aCoder == null) throw new IllegalStateException(
+        "Unable to infer a coder for ValueState and no Coder"
+        + " was specified. Please set a coder by either invoking"
+        + " StateSpecs.value(Coder<T> valueCoder) or by registering the coder in the"
+        + " Pipeline's CoderRegistry.")
+    }
+
+    override def equals(obj: Any): Boolean = {
+      if (obj == this) true
```

Comment:
It's not right in Scala, here without `return`, `true` just becomes a useless statement, which means eventually the last line `Objects.equals(this.aCoder, that.aCoder)` will be executed.